### PR TITLE
AES-NI: use target attributes for x86 32-bit intrinsics

### PR DESCRIPTION
**Pull request by beni-sandu**
## Description

Add target attributes for 32-bit builds to compile with intrinsics out of the box.

This fixes 32-bit compilation issues (e.g. https://github.com/Mbed-TLS/mbedtls/issues/8334 and https://github.com/Mbed-TLS/mbedtls/issues/8400).

For 64-bit, assembly will still be used if cpu flags are not passed when building.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required
- [x] **backport** not required, AES-NI is only enabled on 64bit on 2.28
- [x] **tests** provided

Tested that it builds fine on:
Windows 10: MINGW 32-bit/64-bit (13.0.2), Clang 32-bit/64-bit (17.0.1)
Linux Ubuntu 22.04: gcc 32-bit/64-bit (11.4.0), Clang 32-bit/64-bit (14.0.0)

Let me know if there is some better way of doing this.

### CODE REVIEW

**Comment by tom-daubney-arm** [comment]
Hi @beni-sandu ,

Thanks a lot for the contribution. I have kicked off the CI on your PR so we will get some feedback shortly.

**Comment by tom-daubney-arm** [comment]
Note to maintainers: Internal CI failures are spurious. OpenCI has passed. Marking ready for review.

**Review by daverodgman — CHANGES_REQUESTED** [review]

**Comment by daverodgman** [comment]
It would be good to add a build test for clang and gcc in `all.sh`.

**Comment by beni-sandu** [comment]
Thanks for the feedback @daverodgman, I have addressed your comments.

Seems there was a test for aesni 32-bit intrinsics in all.sh, I have added one for clang too, let me know if that is enough.

**Review by daverodgman — DISMISSED** [review]
LGTM, thanks!

File: library/aesni.h
Comment (tom-cosgrove-arm) [inline library/aesni.h:57]: I don't think we need this comment on the `#endif` - not only is the `#endif` close enough to the `#if` that this just adds noise, the comment isn't (fully) correct. Best to remove it

File: library/aesni.c
Comment (tom-cosgrove-arm) [inline library/aesni.c:55]: While it's nice to have the indentation, the other block added doesn't have it, and nor do (AFAICS) any of the other pre-processor blocks in this file, so best to remove

**Comment by beni-sandu** [comment]
Thanks @tom-cosgrove-arm , fixed now.

**Review by tom-daubney-arm — APPROVED** [review]
LGTM, thanks

**Comment by daverodgman** [comment]
The error is because Ubuntu 16.04 has clang 4.0, which doesn't know about the target attribute. So we'll need a check in `support_test_aesni_m32` in `all.sh` to ensure we have a clang later than 4.0, similar to `support_build_sha_armce`

**Comment by beni-sandu** [comment]
I think it's easier to just check the clang version inside `component_test_aesni_m32`, otherwise skip the clang test?
This way we still run the rest of the aesni tests if we don't have the required version, and we also don't need to add new `component_` and `support_` functions just for the clang test.

Something like this inside `component_test_aesni_m32`:
```
    # If required version of clang is not available, skip this test
    if command -v clang > /dev/null ; then
        # clang >= 4 is required to build with target attributes
        clang_ver="$(clang --version|grep version|sed -E 's#.*version ([0-9]+).*#\1#')"
        if [[ "${clang_ver}" -ge 4 ]] ; then
            # test the intrinsics implementation with clang
            msg "AES tests, test intrinsics (clang)"
            make clean
            make CC=clang CFLAGS='-m32 -Werror -Wall -Wextra' LDFLAGS='-m32'
            # check that we built intrinsics - this should be used by default when supported by the compiler
            ./programs/test/selftest aes | grep "AESNI code" | grep -q "intrinsics"
            grep -q "AES note: using AESNI" ./programs/test/selftest
            grep -q "AES note: built-in implementation." ./programs/test/selftest
            grep -q "AES note: using VIA Padlock" ./programs/test/selftest
            grep -q mbedtls_aesni_has_support ./programs/test/selftest
        fi
    fi
```
I have used >= 4 for clang version, but I am actually not sure if it's okay or needs to be > 4, and cannot double check on my side at the moment.

**Comment by tom-cosgrove-arm** [comment]
At this point, I think you are better off adding new `component_` and `support_` functions for the clang test. They are lightweight shell functions, not a big deal adding new ones. It will help keep the code modular

**Comment by beni-sandu** [comment]
I have added a separate test case for clang, please have a look.

**Review by tom-cosgrove-arm — DISMISSED** [review]
LGTM - thanks

**Comment by tom-cosgrove-arm** [comment]
ping @tom-daubney-arm or @daverodgman for second review?

**Review by daverodgman — DISMISSED** [review]

**Comment by daverodgman** [comment]
@tom-cosgrove-arm I think no changelog because the other PR relating to the common issue #8334 has a changelog?

**Comment by daverodgman** [comment]
Conflicts with #8374

**Comment by beni-sandu** [comment]
Interesting, #8374 mentions that clang does not define `__GNUC__` on Windows, but it did with the versions that I tested, on both Linux/Windows.

I will have another look at this soon.

**Comment by tom-cosgrove-arm** [comment]
> Interesting, #8374 mentions that clang does not define `__GNUC__` on Windows, but it did with the versions that I tested, on both Linux/Windows.

Sorry, probably not very clear.

When clang is used as part of Visual Studio, it is run as the "clang cl driver", `clang-cl.exe`. This is real clang, but like running `clang.exe --driver-mode=cl`, and it behaves like `cl.exe`, not `clang.exe`. This includes defining `_MSC_VER` and not defining `__GNUC__`, but it does define `__clang__`

**Comment by beni-sandu** [comment]
I had a closer look at this, and based on my tests, seems there is some inconsistency between what #8374 is describing in the details and what is actually fixing, unless I am misunderstanding something, or there is a different behaviour in environment because of e.g. versions used.

I understood that one could not compile with clang-cl at all before that patch, with or without CPU flags, but on my side, it was building fine when flags were provided, and throwing an error about missing intrinsics functions when flags are not used. With that patch, it is now throwing an error about missing CPU flags, so it's better.

I anyways reworked my first patch to include those changes and the current behaviour of the development branch in my tests was as follows:

On Windows 10:
----------------------

- Visual Studio 2022, MSVC 19.37.32825:
- [x] 64-bit - Builds with intrinsics out of the box
- [x] 32-bit - Builds with intrinsics out of the box
- Visual Studio 2022, clang-cl 16.0.5:
- [ ] **64-bit - Fails with error on missing CPU flags out of the box,** builds with intrinsics when flags are provided
- [ ] **32-bit - Fails with error on missing CPU flags out of the box,** builds with intrinsics when flags are provided
- Msys2, standalone mingw 13.0.2:
- [x] 64-bit - Builds with assembly out of the box, builds with intrinsics when flags are provided
- [ ] **32-bit - Fails with error on missing CPU flags out of the box,** builds with intrinsics when flags are provided
- Msys2, standalone clang 17.0.1:
- [x] 64-bit - Builds with assembly out of the box, builds with intrinsics when flags are provided
- [ ] **32-bit - Fails with error on missing CPU flags out of the box,** builds with intrinsics when flags are provided


On Linux Ubuntu 22.04:
-------------------------------

- gcc 11.4.0:
- [x] 64-bit - Builds with assembly out of the box, builds with intrinsics when flags are provided
- [ ] **32-bit - Fails with error on missing CPU flags out of the box,** builds with intrinsics when flags are provided
- clang 14.0.0:
- [x] 64-bit - Builds with assembly out of the box, builds with intrinsics when flags are provided
- [ ] **32-bit - Fails with error on missing CPU flags out of the box,** builds with intrinsics when flags are provided

Changes added by the patch in this PR:
==============================

On Windows 10:
----------------------

- Visual Studio 2022, MSVC 19.37.32825:
- [x] 64-bit - Similar to before
- [x] 32-bit - Similar to before
- Visual Studio 2022, clang-cl 16.0.5:
- [x] **64-bit - Now builds with assembly out of the box (instead of throwing error about missing flags), builds with intrinsics when flags are used.**
- [x] **32-bit - Now builds with intrinsics out of the box (instead of throwing error about missing flags)**
- Msys2, standalone mingw 13.0.2:
- [x] 64-bit - Similar to before
- [x] **32-bit - Now builds with intrinsics out of the box (instead of throwing error about missing flags)**
- Msys2, standalone clang 17.0.1:
- [x] 64-bit - Similar to before
- [x] **32-bit - Now builds with intrinsics out of the box (instead of throwing error about missing flags)**


On Linux Ubuntu 22.04:
-------------------------------

- gcc 11.4.0:
- [x] 64-bit - Similar to before
- [x] **32-bit - Now builds with intrinsics out of the box (instead of throwing error about missing flags)**
- clang 14.0.0:
- [x] 64-bit - Similar to before
- [x] **32-bit - Now builds with intrinsics out of the box (instead of throwing error about missing flags)**

Please let me know what you think.

**Review by daverodgman — APPROVED** [review]
Thanks for the thorough testing. LGTM

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Pull request merged**
